### ISSUE TITLE
DataGridMultiSelectColumn | Using The MultiSelectTemplate with a disabled checkbox still triggers selection changed internally

### ISSUE THREAD

**Issue by David-Moreira** [Megabit/Blazorise#5087]
Here's an example of what I'm trying to do, where providing specific markup for the MultiSelectTemplate doesn't work as expected.  Basically, I'd like to have a list of items that, as they are checked, populates another list real time.  Currently, the code I ended up using (diff from below) populates a bound selected list, & when the selections need to be saved, the selected list gets assigned to the data model.  it'd be great if the code could have a master list, where checking & unchecking items populates the data model in real time.  plus, it'd be nice to have the form have a view & edit mode, where the checkboxes can't be checked when in view mode.

```
@page "/"

<Card Shadow="Shadow.Large" Margin="Margin.Is5">
    <CardBody Padding="Padding.Is5">
        <Button Color="Color.Secondary" Clicked="OnChangeEditStateClicked">Change Edit State</Button>
        <br/>
        Edit State = @isEditState
        <br /><br />
        <DataGrid TItem="Foo"
                  Data="@allFoos"
                  PageSize="1000000"
                  FixedHeader="true"
                  Filterable="true"
                  Sortable="true"
                  Editable="@isEditState"
                  SelectionMode="DataGridSelectionMode.Multiple">
            <DataGridColumns>
                <DataGridMultiSelectColumn TItem="Foo" Editable="@isEditState">
                    <MultiSelectTemplate>
                        <Blazorise.Check TValue="bool" Disabled="@(!isEditState)" CheckedChanged="@((isChecked) => { OnCheckChanged(isChecked, context.Item); context.SelectedChanged.InvokeAsync(isChecked); })" />
                    </MultiSelectTemplate>
                </DataGridMultiSelectColumn>
                <DataGridColumn TItem="Foo" Field="@nameof(Foo.Acronym)" Caption="Acronym" />
                <DataGridColumn TItem="Foo" Field="@nameof(Foo.Name)" Caption="Name" />
            </DataGridColumns>
        </DataGrid>
        <br /><br />
        My Foos
        <br />
        <ListView TItem="Foo"
                  Data="myFoos"
                  TextField="(item) => item.Name"
                  ValueField="(item) => item.Acronym"
                  Mode="ListGroupMode.Static"
                  MaxHeight="300px">
        </ListView>
    </CardBody>
</Card>

@code {
    public class Foo
    {
        public Foo()
        {
            Acronym = "";
            Name = "";
        }

        public string? Acronym { get; set; }

        public string? Name { get; set; }
    }

    private List<Foo> allFoos = new List<Foo>
    {
        new Foo { Acronym = "a1", Name = "n1" },
        new Foo { Acronym = "a2", Name = "n2" },
        new Foo { Acronym = "a3", Name = "n3" },
        new Foo { Acronym = "a4", Name = "n4" },
        new Foo { Acronym = "a5", Name = "n5" },
        new Foo { Acronym = "a6", Name = "n6" },
        new Foo { Acronym = "a7", Name = "n7" },
        new Foo { Acronym = "a8", Name = "n8" }
    };

    private List<Foo> myFoos = new List<Foo>();

    private bool isEditState = false;

    private void OnChangeEditStateClicked()
    {
        isEditState = !isEditState;
    }

    private void OnCheckChanged(bool isChecked, Foo foo)
    {
        if (isChecked)
        {
            myFoos.Add(foo);
        }
        else
        {
            myFoos.Remove(foo);    
        }
    }
}
```

_Originally posted by @RealityMasque in https://github.com/Megabit/Blazorise/discussions/5062#discussioncomment-7313790_

**Comment by tesar-tech** [comment]
tl;dr: There is no simple fix and the `_DataGridRowMultipleSelect` is somewhat broken.

The issue here isn't about the disabled checkbox but about `IsSelected` synchronization between the row and the template (the checkbox in this case). 


## Checkbox-Selected Row Sync Issue

In a "normal case" of multi-select functionality, it appears that synchronization is working, but that's just because the row is being clicked at the same time. This causes both events (selected row and checked checkbox) to be in sync.

One can simply replace the whole `Blazorise.Check` inside `Label`, along with all the event callbacks, in `_DataGridRowMultipleSelect.razor`:

```
<Label @onclick="@OnCheckedClicked">
    <Check TValue="bool" Checked="@Checked" CheckedChanged="@OnCheckedChanged"
           CheckedExpression="@(() => Checked)"></Check>
</Label>
```

with:

```
<input type="checkbox" checked="@Checked"/>
```

and it will work almost the same. Almost—because it doesn't work with `RowSelectableHandler` and `DataGridSelectReason`. Also, for this simple case, it actually works better...

v1.7.4 : Selecting a row results in the inability to de-select it using the checkbox:  

https://github.com/user-attachments/assets/2c9ac1ce-ab5c-4e28-8fc4-127fb23e7495  

Replaced with `<input checkbox>`:  

https://github.com/user-attachments/assets/d2208629-91dc-4c38-b3a1-d5b044bcf925  

(There is more to this bug, but it's not important for now.)  

This issue extends deeper into `_DataGridRowMultipleSelect.razor.cs` and its clicked and checked events. It would really deserve some refactoring.



Also the following code from the provided example:

```
<Blazorise.Check TValue="bool" Disabled="@(!isEditState)" CheckedChanged="@((isChecked) => { OnCheckChanged(isChecked, context.Item); context.SelectedChanged.InvokeAsync(isChecked); })" />
```

can easily be simplified by removing the `SelectedChanged` invocation:

```
<Blazorise.Check TValue="bool" Disabled="@(!isEditState)" CheckedChanged="@((isChecked) => { OnCheckChanged(isChecked, context.Item); })"/>
```

This is important because, without proper synchronization between the "built-in checkbox" and the selected row, we cannot ensure it will work with a "templated checkbox."


## MultiSelectTemplate

`MultiSelectTemplate` is meant for customizing the checkbox. However, customization can also be done without anything else:

```
<MultiSelectTemplate>
  <div>
    @(context.IsSelected ? "\u2705" : "\u274c")
  </div>  
</MultiSelectTemplate>
```

Resulting in:  

https://github.com/user-attachments/assets/8bf5f901-cd4b-41db-8bc2-191daa80fa62  

This works as expected because the selected state is transferred to the `div` with emojis.

The actions in the `div` have nothing to do with the selection mechanism. The same applies to `Blazorise.Check` inside `MultiSelectTemplate`. There is no mechanism that detects something being disabled and turns off the selection process. The only existing option is `PreventRowClicked`, as David suggested, which works fine for this purpose.


## Solution

Implementing a synchronization mechanism to connect `datagridRow.IsSelectable` with `MultiSelectTemplate.SomeStufInThere.IsDisabled` is possible, but for the user, it would likely be nearly the same as implementing David’s fix.

Instead, I would suggest mentioning the "de-sync" nature of `MultiSelectTemplate` in the documentation and fixing the related bugs.

@stsrki ?

**Comment by stsrki** [comment]
First of all, I would like to move this to 2.0. Whatever we touch in 1.7, I can already see that bringing it up to 2.0 would be problematic due to git conflicts.

---

Regarding the change from 

```
<Label @onclick="@OnCheckedClicked">
    <Check TValue="bool" Checked="@Checked" CheckedChanged="@OnCheckedChanged"
           CheckedExpression="@(() => Checked)"></Check>
</Label>
```

to

```
<input type="checkbox" checked="@Checked"/>
```

We cannot do that. Leaving the regular `input` will break any CSS styling for our providers(bootstrap, tailwind, etc). We have to use the Check component. 

Also, I was never a fan of wrapping `Check` with `Label`. The Check internally already has a label and an input with a HTML structure for the particular CSS provider. If possible, I would replace Label wrapper with something else that is not going to affect the Check lifecycle. Maybe a Div?

---

As for the other changes you mentioned. TBH, I don't quite follow it, as I would have to invest some time to learn about the lifecycle and event callbacks and how all of them play together. But, in any case, if you think it needs to change, change it. I will review it in PR later on.

**Comment by tesar-tech** [comment]
> We cannot do that.

Oh yeah, I wasn't suggesting such change, I was just demonstrating the functionality and the issue...

>  I would replace Label wrapper with something else 

Yes, the Label also confuses me with the click event. From my perspective it just adds another (unnecessary)  event-chain into the mix...

I will look into that more and see what's possible.
